pipeline/pipeline/task.py

```diff
@@ -40,12 +53,20 @@ def __getitem__(self, item):
     def __getattr__(self, item):
         return self._get_resource(item)
 
+    def _add_outputs(self, resource):
```

Comment:
Do we need this and _add_inputs, or would it be sufficient for methods calling _add_outputs / _add_inputs to call (for instance) `_add_resource_to_set(self._outputs)`. The calling method already abstracts away a private method; it kind of seems we now abstract the same private method twice.

Reply:
Your way is probably better. Let me try that.

Reply:
The problem with your proposal is it works fine when mutating your own data structures, but I also want to mutate another Task's data structure as well. Then it would change from this `r._source._add_outputs(r)` to `_add_resource_to_set(r._source._outputs, r)` which I felt was a bit weird. The right thing to do is to not have the helper method, but I didn't want to copy the same logic for the resource group twice.